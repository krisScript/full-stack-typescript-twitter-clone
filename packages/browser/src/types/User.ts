import CommonUser from '@twtr/common/source/types/user';

export default interface User extends CommonUser {
  following: string[];
  likes: string[];
  bookmarks: string[];
  replies: string[];
  retweets: string[];
  avatar?: string;
  _id: string;
}
