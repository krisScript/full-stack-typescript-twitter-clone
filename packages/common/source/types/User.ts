export default interface User {
  username: string;
  handle: string;
  email: string;
  website: string | undefined;
  profilePhoto: string;
  headerPhoto: string;
  date: string;
  followers: number;
}
