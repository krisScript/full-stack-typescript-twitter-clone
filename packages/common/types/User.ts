export default interface User {
  username: string;
  handle: string;
  email: string;
  website: string | undefined;
  date: string;
}
