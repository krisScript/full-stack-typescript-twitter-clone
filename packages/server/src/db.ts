import mongoose from "mongoose";

const mongoURI: string = `mongodb+srv://${process.env.MONGO_USER}:${process.env.MONGO_PASSWORD}@cluster0-zmcyw.mongodb.net/${process.env.MONGO_DATABASE}?retryWrites=true`;
const db = async (): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, { useNewUrlParser: true });
  } catch (err) {}
};

export default db;
