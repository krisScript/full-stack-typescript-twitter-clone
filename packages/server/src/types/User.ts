import CommonUser from '@twtr/common/source/types/User';
import { Schema, Document } from 'mongoose';

export default interface User extends CommonUser, Document {
  password: string;
  confirmed: boolean;
  following: Schema.Types.ObjectId[];
  likes: Schema.Types.ObjectId[];
  bookmarks: Schema.Types.ObjectId[];
}
