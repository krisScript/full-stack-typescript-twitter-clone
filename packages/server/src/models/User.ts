import mongoose, { Schema } from 'mongoose';
import User from '@customTypes/User';

const UserSchema: Schema = new Schema({
  username: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  handle: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  website: { required: false, type: String, minlength: 3 },
  confirmed: { type: Boolean, default: false },
  profilePhoto: { type: String },
  headerPhoto: { type: String },
  date: {
    type: Date,
    default: Date.now,
  },
  following: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  followers: {
    type: Number,
    default: 0,
  },
  likes: [
    {
      source: {
        type: mongoose.Types.ObjectId,
        required: true,
        refPath: 'likes.ref',
      },
      ref: {
        type: String,
        required: true,
        enum: ['Tweet', 'Reply'],
      },
    },
  ],
  bookmarks: [
    {
      source: {
        type: mongoose.Types.ObjectId,
        required: true,
        refPath: 'bookmarks.ref',
      },
      ref: {
        type: String,
        required: true,
        enum: ['Tweet', 'Reply'],
      },
    },
  ],
});

export default mongoose.model<User>('User', UserSchema);
