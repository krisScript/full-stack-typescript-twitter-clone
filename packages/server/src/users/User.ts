import mongoose, { Schema } from 'mongoose';
import User from '../types/User';
import duplicationErrorHandler from '../middleware/duplicationErrorHandler';
//@ts-ignore
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  handle: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  isConfirmed: { type: Boolean, default: true },
  avatar: {
    type: String,
  },
  cover: { type: String },
  date: {
    type: Date,
    default: Date.now,
  },
  following: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  followers: {
    type: Number,
    default: 0,
  },
  likes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
  retweets: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
  replies: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],

  bookmarks: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
});

UserSchema.index({ handle: 'text' });

UserSchema.plugin(uniqueValidator);

export default mongoose.model<User>('User', UserSchema);
