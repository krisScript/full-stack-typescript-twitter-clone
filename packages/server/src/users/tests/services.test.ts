import mongoose from 'mongoose';
import jwt from 'jsonwebtoken';
import {
  areCredentialsAvailable,
  getUserById,
  getUserByEmail,
} from 'src/users/services';
import User from 'src/users/User';
import db from 'src/db';
import getResource from '@utilities/getResource';

jest.mock('@utilities/getResource');

const getResourceMock = getResource as jest.MockedFunction<typeof getResource>;

jest.mock('jsonwebtoken');

const mockToken = 'mockToken';
(jwt.sign as jest.Mock).mockReturnValue(mockToken);

describe('userServices', (): void => {
  beforeAll(
    async (): Promise<void> => {
      db();
      await User.deleteMany({}).exec();
    },
  );
  afterEach(
    async (): Promise<void> => {
      await User.deleteMany({}).exec();
    },
  );
  afterAll(
    async (): Promise<void> => {
      await mongoose.disconnect();
    },
  );
  const username = 'username';
  const handle = 'testUserHandle';
  const email = 'testmail@mail.com';
  const password = 'testPassword';
  describe('getUserByEmail', (): void => {
    it(`should call getResource`, async (): Promise<void> => {
      expect.assertions(2);
      await getUserByEmail(email);

      expect(getResource).toHaveBeenCalledTimes(1);
      expect(getResourceMock).toHaveBeenCalledWith(User, {
        name: 'email',
        value: email,
      });
    });
  });
  describe('getUserById', (): void => {
    it(`should call getResource`, async (): Promise<void> => {
      expect.assertions(2);
      const userId = 'userId';
      await getUserById(userId);

      expect(getResource).toHaveBeenCalledTimes(1);
      expect(getResourceMock).toHaveBeenCalledWith(
        User,
        {
          name: '_id',
          value: userId,
        },
        '',
      );
    });
  });

  describe('areCredentialsAvailable', (): void => {
    it('should throw an error when the user credentials are already taken', async (): Promise<
      void
    > => {
      expect.assertions(2);
      const credentials: {
        path: 'username' | 'handle' | 'email';
        value: string;
      }[] = [
        { path: 'username', value: username },
        { path: 'handle', value: handle },
        { path: 'email', value: email },
      ];
      await User.insertMany({ username, handle, email, password });
      await expect(areCredentialsAvailable(credentials)).rejects.toThrow();
      await expect(
        areCredentialsAvailable(credentials),
      ).rejects.toMatchSnapshot();
    });
    it("shouldn't throw an error when the user credentials are available", async (): Promise<
      void
    > => {
      const credentials: {
        path: 'username' | 'handle' | 'email';
        value: string;
      }[] = [
        { path: 'username', value: username },
        { path: 'handle', value: handle },
        { path: 'email', value: email },
      ];
      expect.assertions(1);
      await expect(
        areCredentialsAvailable(credentials),
      ).resolves.toBeUndefined();
    });
  });
});
