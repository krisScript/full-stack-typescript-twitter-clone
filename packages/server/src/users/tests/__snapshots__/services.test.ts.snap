// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userServices areCredentialsAvailable should throw an error when the user credentials are already taken 1`] = `[Error: Conflict]`;
